src/Microsoft.AspNetCore.Sockets.Common/BinaryMessageFormatter.cs

```diff
@@ -0,0 +1,130 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Binary;
+using System.IO.Pipelines;
+
+namespace Microsoft.AspNetCore.Sockets
+{
+    internal static class BinaryMessageFormatter
+    {
+        private const byte TextTypeFlag = 0x00;
+        private const byte BinaryTypeFlag = 0x01;
+        private const byte ErrorTypeFlag = 0x02;
+        private const byte CloseTypeFlag = 0x03;
+
+        internal static bool TryFormatMessage(Message message, Span<byte> buffer, out int bytesWritten)
+        {
+            // We can check the size needed right up front!
+            var sizeNeeded = sizeof(long) + 1 + message.Payload.Buffer.Length;
+            if (buffer.Length < sizeNeeded)
+            {
+                bytesWritten = 0;
+                return false;
+            }
+
+            buffer.WriteBigEndian((long)message.Payload.Buffer.Length);
+            if (!TryFormatType(message.Type, buffer.Slice(sizeof(long), 1)))
+            {
+                bytesWritten = 0;
+                return false;
+            }
+
+            buffer = buffer.Slice(sizeof(long) + 1);
+
+            message.Payload.Buffer.CopyTo(buffer);
+            bytesWritten = sizeNeeded;
+            return true;
+        }
+
+        internal static bool TryParseMessage(ReadOnlySpan<byte> buffer, out Message message, out int bytesConsumed)
+        {
+            // Check if we have enough to read the size and type flag
+            if (buffer.Length < sizeof(long) + 1)
+            {
+                message = default(Message);
+                bytesConsumed = 0;
+                return false;
+            }
+
+            // REVIEW: The spec calls for 64-bit length but I'm thinking that's a little ridiculous.
```

Comment:
The biggest thing is that absolutely **nothing** in our stack can handle a 64-bit length buffer. We'll need to build a bunch of new primitives (buffer chains, etc.) to handle it. And a 2GB-long **single message** is really ridiculous. If someone wants to upload files over 2GB via this, they just have to split it into separate messages, which they **really should do already**.